### PR TITLE
[MRG] Feature: email alerts

### PR DESCRIPTION
**Pull request by aapeliv**
Continuation from #212 and #263 as Sam's busy these days.

### CODE REVIEW

**Comment by aapeliv** [comment]
This is ready, but I feel like it's missing something.

**Comment by aapeliv** [comment]
Few things come to mind:
* Should emails probably be sent asynchronously with some kind of queueing system?
* This is probably prone to injection, as there's a bunch of user-defined fields, some sanitisation should be happening somewhere?
* Similarly, is there some kind of max limit on names, etc?
* Emails for new messages are not actually sent anywhere, and currently the emails this prepares are annoying as they don't mention the message sender or content.

File: app/backend/src/couchers/tasks.py [hunk 1]
```diff
@@ -28,26 +29,70 @@ def send_login_email(user, token, expiry_text):
...
-            "username_reported": username_reported,
-            "reason": reason,
-            "description": description,
+            "complaint": complaint,
+        },
+    )
+
+
+def send_host_request_email(host_request):
+    host_request_link = f"{config['BASE_URL']}/hostrequests/"
```
Comment (lucaslcode) [inline app/backend/src/couchers/tasks.py]: I wonder if it would be good to have these links as constants in one file.

**Review by lucaslcode — DISMISSED** [review]

**Comment by lucaslcode** [comment]
> Should emails probably be sent asynchronously with some kind of queueing system?

Is the send email function blocking? 

> This is probably prone to injection, as there's a bunch of user-defined fields, some sanitisation should be happening somewhere?

It all goes through the templating where it sanitizes html, no?

> Similarly, is there some kind of max limit on names, etc?

Not from memory...

> Emails for new messages are not actually sent anywhere, and currently the emails this prepares are annoying as they don't mention the message sender or content.

Yes

**Comment by aapeliv** [comment]
> Is the send email function blocking?

Yep. At the moment it's pretty fast since it's calling AWS SES in the same AWS region.


> It all goes through the templating where it sanitizes html, no?

It seems you're right, jinja escapes HTML to prevent XSS.


> > Emails for new messages are not actually sent anywhere, and currently the emails this prepares are annoying as they don't mention the message sender or content.
> 
> Yes

I feel we should implement an async task runner, and then as a starting point only send an email if you haven't seen the message within 15 minutes or something.

**Comment by kalvdans** [comment]
> Yep. At the moment it's pretty fast since it's calling AWS SES in the same AWS region.

I'm pretty fluent in exim4 configuration, we could use it as smarthost Mail Transportation Agent with its own mail queue and retry loop.

File: app/backend/src/tests/test_email.py [hunk 2]
```diff
@@ -124,3 +95,113 @@ def mock_send_email(sender_name, sender_email, recipient, subject, plain, html):
...
+            assert to_date in plain
+            assert to_date in html
+            assert from_user.avatar_url not in plain
+            assert from_user.avatar_url in html
+            assert f"{config['BASE_URL']}/hostrequests/" in plain
+            assert f"{config['BASE_URL']}/hostrequests/" in html
+            return message_id
+
+        with patch("couchers.email.send_email", mock_send_email):
+            send_host_request_email(host_request)
```
Comment (kalvdans) [inline app/backend/src/tests/test_email.py:153]: verify that our mock function was called once.

File: app/backend/src/tests/test_email.py [hunk 3]
```diff
@@ -124,3 +95,113 @@ def mock_send_email(sender_name, sender_email, recipient, subject, plain, html):
...
+            assert from_user.name in plain
+            assert from_user.name in html
+            assert from_user.avatar_url not in plain
+            assert from_user.avatar_url in html
+            assert f"{config['BASE_URL']}/friends/" in plain
+            assert f"{config['BASE_URL']}/friends/" in html
+            return message_id
+
+        with patch("couchers.email.send_email", mock_send_email):
+            send_friend_request_email(friend_relationship)
```
Comment (kalvdans) [inline app/backend/src/tests/test_email.py:200]: ditto.

**Review by kalvdans — DISMISSED** [review]
just two minor remarks on the tests.

File: app/backend/src/tests/test_email.py (hunk 2 above)
Comment (aapeliv) [reply app/backend/src/tests/test_email.py:153]: Good idea

File: app/backend/src/couchers/tasks.py (hunk 1 above)
Comment (aapeliv) [reply app/backend/src/couchers/tasks.py]: Good idea.

**Comment by aapeliv** [comment]
@lucaslcode: added `couchers.urls` for having all the URL specs in the same file

@kalvdans: added checks that the calls are called exactly once.

**Comment by aapeliv** [comment]
> > Yep. At the moment it's pretty fast since it's calling AWS SES in the same AWS region.
> 
> I'm pretty fluent in exim4 configuration, we could use it as smarthost Mail Transportation Agent with its own mail queue and retry loop.

I think for now it's OK to merge this synchronous email sending. But I've created a new issue #282 to talk about background jobs.

File: app/backend/src/tests/test_email.py (hunk 2 above)
Comment (kalvdans) [reply app/backend/src/tests/test_email.py:153]: Cool, didn't know about create_autospec before :)

**Review by kalvdans — APPROVED** [review]
Looks fine to me!

**Pull request merged**